filewatcher/filewatcher.go

```diff
@@ -15,270 +15,180 @@
-			errors: make(chan error, 1),
-			md5Sum: md5Sum,
-		}
+	return nil
+}
 
-		return nil
-	}
+// Add a path to watch
+func (fw *fileWatcher) Add(path string) error {
+	fw.mu.Lock()
+	defer fw.mu.Unlock()
 
-	// Build a new worker if not exist.
-	watcher, err := fsnotify.NewWatcher()
+	ws, cleanedPath, _, err := fw.getWorker(path)
 	if err != nil {
 		return err
 	}
-	if err = watcher.Add(parentPath); err != nil {
-		watcher.Close()
-		return err
-	}
-
-	var md5Sum string
-
-	if _, err = os.Stat(cleanedPath); err == nil {
-		md5Sum, err = getMd5Sum(cleanedPath)
-		if err != nil {
-			watcher.Close()
-			return fmt.Errorf("failed to get md5 sum for %s: %v", path, err)
-		}
-	} else if !os.IsNotExist(err) {
-		// if the file doesn't exist, ignore the md5 sum.
-		watcher.Close()
-		return fmt.Errorf("failed to stat file: %s: %v", path, err)
-	}
 
-	wk := &worker{
-		watcher: watcher,
-		watchedFiles: map[string]*fileTracker{
-			cleanedPath: {
-				events: make(chan fsnotify.Event, 1),
-				errors: make(chan error, 1),
-				md5Sum: md5Sum,
-			},
-		},
-	}
-	w.workers[parentPath] = wk
-
-	go func() {
-		for {
-			select {
-			case event, ok := <-watcher.Events:
-				if !ok { // 'Events' channel is closed
-					return
-				}
-
-				for path, tracker := range wk.watchedFiles {
-					newSum, _ := getMd5Sum(path)
-					if newSum != "" && newSum != tracker.md5Sum {
-						tracker.md5Sum = newSum
-						tracker.events <- event
-						break
-					}
-				}
-			case err, ok := <-watcher.Errors:
-				if !ok {
-					// 'Errors' channel is closed. Construct an error for it.
-					// We'll only close worker errors channel in "Remove" for consistency.
-					err = errors.New("channel closed")
-				}
-				for _, tracker := range wk.watchedFiles {
-					tracker.errors <- err
-				}
-				return
-			}
-		}
-	}()
+	ws.worker.addPath(cleanedPath)
+	ws.count++
 
 	return nil
 }
 
-// Remove is implementation of Remove interface of FileWatcher.
-func (w *fsNotifyWatcher) Remove(path string) error {
-	w.mu.Lock()
-	defer w.mu.Unlock()
-
-	return w.remove(path)
-}
-
-func (w *fsNotifyWatcher) remove(path string) error {
-	cleanedPath, parentPath := formatPath(path)
+// Stop watching a path
+func (fw *fileWatcher) Remove(path string) error {
+	fw.mu.Lock()
+	defer fw.mu.Unlock()
 
-	worker, workerExist := w.workers[parentPath]
-	if !workerExist {
-		return fmt.Errorf("path %s already removed", path)
+	ws, cleanedPath, parentPath, err := fw.getWorker(path)
+	if err != nil {
+		return err
 	}
 
-	tracker, pathExist := worker.watchedFiles[cleanedPath]
-	if !pathExist {
-		return fmt.Errorf("path %s already removed", path)
-	}
+	ws.worker.removePath(cleanedPath)
 
-	delete(worker.watchedFiles, cleanedPath)
-	close(tracker.events)
-	close(tracker.errors)
-	if len(worker.watchedFiles) == 0 {
-		// Remove the watch if all of its paths have been removed.
-		delete(w.workers, parentPath)
-		return worker.watcher.Close()
+	ws.count--
+	if ws.count == 0 {
```

Comment:
If you embed count operations into removePath/addPath (and perhaps make removePath return bool), then you can avoid changing a member field on worker, and avoid workerState struct.